text/0074-source-context-via-links.md

```diff
@@ -0,0 +1,70 @@
+- `post_context` - A list of source code lines after `context_line` (in order) – usually [`lineno + 1:lineno + 5`].
+
+Showing a source code context and a link to it (e.g. to GitHub) currently works only if the source code itself is
+available during symbolication, thus the attributes are filled in at that time, or if they were already sent with the event by the SDK.
+
+There are, however, situations where we do have a URL where the source code resides, but not the contents (without downloading it), for example:
+
+- Portable-PDB source-link (.NET, see [this `symbolic` issue](https://github.com/getsentry/symbolic/issues/735))
+- via a repository integration in combination with associated commit
+
+In these cases, the actual source code is not necessary to do symbolication (as opposed to source maps and other types of obfuscated source containers) but it is useful for end-user
+when evaluating the issue in the UI.
+
+## Related GH feature requests
+
+- [Extend stack-trace linking UI](https://github.com/getsentry/sentry/issues/35608)
+- [Support stack trace linking when stack frames do not contain line context.](https://github.com/getsentry/sentry/issues/44015)
+
+# Options Considered
+
+## A | Download the source code during symbolication
+
+One option is to download the source code from its link at the time the event is symbolicated, filling the context-related
+`Frame` attributes for each stacktrace frame, while the event is being processed.
+
+### Pros
+
+- Reuses existing fields (changes only needed in the `Symbolicator`).
+- Sources surrounding the frame line are then stored in the database, enabling [search-by-code](https://github.com/getsentry/sentry/issues/3755).
+- May be possible to do server-side authentication (based on project/org configuration) - this would complicate the solution & caching though, as opposed to only supporting publicly available sources.
+
+### Cons
+
+- Downloads all frame sources even though the event/frame may never be viewed by a user.
+- Adds potentially tens of requests for each symbolicated event that has a source link - as many as there are in the stack trace.
+
+## B | Add a source-link to the `Frame` and let the UI handle it
+
+Another option is to extend `Frame` attributes with a new field, e.g. `source_link` and let UI download the source code
+as needed or just display the link.
+
+### Pros
+
+- No overhead on the server for loading sources in situations where the event/frame wouldn't be shown.
+- Requests to fetch the source code, if any, are made by the user browser, thus avoiding potential quota limits.
+
+### Cons
+
+- Sources not available in the database -> can't [search-by-code](https://github.com/getsentry/sentry/issues/3755).
+- Accessing private sources may be problematic. We could still show the source link, though.
+
+# Unresolved questions
+
+- Neither approach is a clear winner, any suggestions?

```

Comment:
I believe from an implementation POV, we can even do a hybrid approach:
- Always attach a source link to stack frames.
- Only download/apply source context for the top N frames of the crashing thread.

With that, we can focus the ahead-of-time work to the context the end user is most likely to actually look at, and search for; while keeping options open for the future.

I believe this is big enough to involve more people from the product side of things, to answer the following questions:

- Which parts of the product need source context?
- How much (lines of) source context?
- How does this tie into search-by-code?
- Do we need *all* the source context at once, or is it okay to lazily load some of it at a later time (favoring a hybrid approach)